layer/layer_store.go

```diff
@@ -563,7 +563,7 @@ func (ls *layerStore) ReleaseRWLayer(l RWLayer) ([]Metadata, error) {
 	}
 
 	if m.hasReferences() {
-		return []Metadata{}, nil
+		return []Metadata{}, ErrActiveMount
```

Comment:
Yes, I agree with you. But when we delete a container, if the rwlayer is still busy, we should not do the `defer func()` in #L116, that is to say, we need an error return from `daemon.layerStore.ReleaseRWLayer`.
```go
 80 func (daemon *Daemon) cleanupContainer(container *container.Container, forceRemove, removeVolume bool) (err error) {
...
114         // If force removal is required, delete container from various
115         // indexes even if removal failed.
116         defer func() {
117                 if err == nil || forceRemove {
118                         daemon.nameIndex.Delete(container.ID)
119                         daemon.linkIndex.delete(container)
120                         selinuxFreeLxcContexts(container.ProcessLabel)
121                         daemon.idIndex.Delete(container.ID)
122                         daemon.containers.Delete(container.ID)
123                         if e := daemon.removeMountPoints(container, removeVolume); e != nil {
124                                 logrus.Error(e)
125                         }
126                         daemon.LogContainerEvent(container, "destroy")
127                 }
128         }()
129
130         if err = os.RemoveAll(container.Root); err != nil {
131                 return fmt.Errorf("Unable to remove filesystem for %v: %v", container.ID, err)
132         }
133
134         // When container creation fails and `RWLayer` has not been created yet, we
135         // do not call `ReleaseRWLayer`
136         if container.RWLayer != nil {
137                 metadata, err := daemon.layerStore.ReleaseRWLayer(container.RWLayer)
138                 layer.LogReleaseMetadata(metadata)
139                 if err != nil && err != layer.ErrMountDoesNotExist {
140                         if err == layer.ErrActiveMount {
141                                 return fmt.Errorf("Sorry, the rwlayer of container %v is busy right now", container.ID)
142                         }
143                         return fmt.Errorf("Driver %s failed to remove root filesystem %s: %s", daemon.GraphDriverName(), container.ID, err)
144                 }
145         }
146
147         return nil
148 }
```